virttest/qemu_devices/qcontainer.py

```diff
@@ -4031,9 +4031,41 @@ def hostdev_define_by_params(self, name, params, bus=None):
                 "failover_pair_id": params.get("vm_hostdev_failover_pair_id"),
             }
         )
+
+        devs = []
+
+        iommufd_id = params.objects("vm_hostdev_iommufd")
```

Comment:
I suppose this is for all vfio devices, but each device has its namespace. like I can set `vm_hostdev_iommufd_hostdev1 = iommufd0` and `vm_hostdev_iommufd_hostdev2 = iommufd1`. As we know, every vfio device can only use 1 iommufd object, so let this parameter be used for device namespace can avoid using `break`.
```python
iommufd_id = params.get("vm_hostdev_iommufd")
if iommufd_id:
    dev_params["iommufd"] = iommufd_id
    if not self.get_by_qid(iommufd_id):
        iommufd = self.iommufd_object_define_by_params(iommufd_id)
        devs.append(iommufd)
```

But if you have other requirements like negative test, please let me know.

Reply:
@PaulYuuu 
This break is just for vm_hostdev_iommufd_hostdev1 = iommufd0 and vm_hostdev_iommufd_hostdev2 = iommufd1.
Because this function hostdev_define_by_params will be called by each device, for hostdev2, it will be called again.
But we use "for" in this function to check all iommufd id we set in vfio_net_boot.cfg in tp_qemu:
```
        # multi hostdev will be bound to different iommufd
        - multi_iommufd:
            vm_hostdev_iommufd = iommufd0 iommufd1
```
So without break, for hostdev1, it will be set to vm_hostdev_iommufd_hostdev1 = iommufd0 in first iteration, and then be overwritten to vm_hostdev_iommufd_hostdev1 = iommufd1 in the next interation.
Then the para will be like:
vm_hostdev_iommufd_hostdev1 = iommufd1(iommufd0 will be overwritten in this loop)
vm_hostdev_iommufd_hostdev2 = iommufd1

And we cannot map iommufd outside hostdev_define_by_params function
So, I just think we can use "break" here, I have tested it, the para can be set as we expected.

Reply:
If I want to test 2 vfio devices, 1 with iommufd, 1 without, for the current design, this is impossible, causes the only one iommufd object will attend to all vfio devices.

> So without break, for hostdev1, it will be set to vm_hostdev_iommufd_hostdev1 = iommufd0 in first iteration, and then be overwritten to vm_hostdev_iommufd_hostdev1 = iommufd1 in the next interation.
Then the para will be like:
vm_hostdev_iommufd_hostdev1 = iommufd1(iommufd0 will be overwritten in this loop)
vm_hostdev_iommufd_hostdev2 = iommufd1

As I said, vfio devices have their namespace during the device creation. in qemu_vm.py
```python
        for hostdev in params.objects("vm_hostdevs"):
            hostdev_params = params.object_params(hostdev)
            dev = devices.hostdev_define_by_params(hostdev, hostdev_params)
```
`object_params` will cut out `vm_hostdev_iommufd_hostdev1 = iommufd0` to `vm_hostdev_iommufd = iommufd0` before `hostdev_define_by_params`. When creating hostdev2,  `vm_hostdev_iommufd_hostdev2 = iommufd1` will be `vm_hostdev_iommufd = iommufd1`, so `hostdev_define_by_params` can always handle the correct iommufd definition. And if you want to test all vfio with same iommufd, just set `vm_hostdev_iommufd = iommufd0`.

```
# with same iommufd
vm_hostdev_iommufd = iommufd0

# with diff iommufd
vm_hostdev_iommufd_hostdev1 = iommufd0
vm_hostdev_iommufd_hostdev2 = iommufd1
```

You can read the context from qemu_vm and also check `object_params` func.

Reply:
Thank you @PaulYuuu ,to double confirm, do you mean we can fix the iommufd id for each device in vfio_net_boot.cfg like this?
# with diff iommufd
vm_hostdev_iommufd_hostdev1 = iommufd0
vm_hostdev_iommufd_hostdev2 = iommufd1

Reply:
That's right!

Reply:
@PaulYuuu Thanks for your advice, updated this PR, and updated https://github.com/autotest/tp-qemu/pull/4163 as well, could you help to check again?